Rewrite multiplayer to not inherit `RoomSubScreen`

smoogipoo opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/31882

In a similar fashion to #31882, this is a full rewrite of `MultiplayerMatchSubScreen` that no longer inherits (and finally removes) `RoomSubScreen`. The primary goals here are:
- For beatmap/ruleset/mod bindables to be removed from the screen.
- For "selected item" bindable to be removed. This requirement happened in addition to everything else, because leaving it in increases the number of mental hoops you need to reason around states.
- For, at the end of the day, `MultiplayerRoom` and related models to be authoritative over the client's state. In general, the expectation is for client-local states to only update after a round-trip to the server.

Unlike the playlists PR, this one needs to go a little bit further, and so includes reimplementations of:
- Many components that previously relied on the item/beatmap/ruleset/mods bindables.
- The user mod select overlay. This one is a bit of a "special" component and so features logic you'll not find anywhere else, that deals with both client-authoritative (e.g. user quickly toggles a mod on/off) and server-authoritative (e.g. ruleset is changed) flows.

smoogipoo committed:
Add initial implementation of new match screen

smoogipoo committed:
Reimplement mod display to not use bindable user mods

smoogipoo committed:
Reimplement user mod selection to not use bindable user mods

smoogipoo committed:
Fix screen being exited before confirmed

smoogipoo committed:
Handle presented beatmaps

smoogipoo committed:
Adjust setup to use MultiplayerRoom

smoogipoo committed:
Fix initial title

smoogipoo committed:
Replace existing screen with new implementation

Fixing some tests/making them compile in the process.

smoogipoo committed:
Remove RoomSubScreen, update tests further

smoogipoo committed:
Merge branch 'room-subscreen-split' into multi-subscreen-split

smoogipoo committed:
Resolve simple test failures

smoogipoo committed:
Block screen input during gameplay

smoogipoo committed:
Merge branch 'room-subscreen-split' into multi-subscreen-split

smoogipoo committed:
Hide mod select overlay when leaving screen

smoogipoo committed:
Fix mod select overlay active mods not recomputed

smoogipoo committed:
Add forgotten partial

smoogipoo committed:
Merge branch 'master' into multi-subscreen-split

smoogipoo committed:
Apply stylistic changes

bdach commented:
Is this still blocked? I guess compile is failing but that seems trivially resolvable.

smoogipoo committed:
Remove "CurrentItem" from match subscreen

I really didn't like this, because I have nightmares of the playlists
subscreens' `SelectedItem`. The way it works here is really weird too,
because it includes the user's style for "completeness" reasons, which
makes it difficult to understand whether components will behave
correctly.

All components that previously used a `Bindable<PlaylistItem>` have been
updated to use `MultiplayerClient` events. It's not quite the prettiest
change, and may take a few revisions to look pretty.

smoogipoo commented:
Yes, this is still a WIP with significant changes atm. Will re-draft.

smoogipoo committed:
Simplify with helper property

smoogipoo committed:
Fix crash

smoogipoo committed:
Fix incorrect condition, simplify gameplay start logic

The condition in question was the `|| Beatmap.IsDefault` in
`onLoadRequested()`, which should've been an `&&` instead.

The code is pretty jank though, so I've opted instead to merge the two
methods and remove the extra validation logic. Beatmap availability
tracker takes on the role of ensuring gameplay can be started in the
first place.
Other updates (e.g. `SettingsChanged`/`UserStyleChanged`) are expected
to/will always arrive from the server prior to `LoadRequested`, so the
extra validation here is really unnecessary.

We can add a `updateGameplayState()` if desired, though it should be
unnecessary and not added to keep things simple. If this does fail, I'll
earlier question my thinking process in this PR.

smoogipoo committed:
Remove unnecessary validation

smoogipoo committed:
Fix current user style not displayed correctly

smoogipoo committed:
Remove another "selected item" bindable

smoogipoo committed:
Safeguard against potential multiple adds

Just in case - don't know if this is possible in practice - I've long
since forgotten if the historical items have their playlist order
updated.

smoogipoo committed:
Fix CI inspection

smoogipoo commented:
I'm tentatively unblock and marking this as ready for review.

I've done a self-review pass and some testing, and am generally okay with what it's ended up as. Have updated the description to be more accurate with the current state of things.

Tentative because I expect this to receive quite a bit of friction with so many components moving, and I would not be surprised to expect additional tests to be created, and I'll probably continue making those myself as time allows.

smoogipoo commented:
Blocking this one again for now because I forgot to take a look back at scheduling. Most of these callbacks likely should be scheduled in general so that components aren't updating in the background during gameplay.

Review can still continue as I don't expect changes to be significant, and are not required for proper operation.

smoogipoo committed:
Fix items not updating in queue

smoogipoo committed:
Schedule callbacks where possibly important

These are just places where performance could be impacted if an update
arrived during gameplay, where correctness is not affected.

Ensuring correctness (as opposed to not impacting it) is less of a
concern it is expected to be handled elsewhere, and most correct is to
not have _any_ schedules at all.
For instance, you could theorise a situation where the queue mode is
changed during gameplay, and then the current playlist item is changed
due to reordering, and come to reason that `updateGameplayState()`
should be scheduled. This would be an example of trying to "ensure
correctness" via these schedules, but in practice this can't happen
because the server blocks changes to the room's settings during
gameplay.

That is why the only consideration is for these changes is that they do
not _impact_ existing correctness.

smoogipoo commented:
With https://github.com/ppy/osu/pull/32250/commits/3a94f4d5c2a0acf2e60d368cb21ead41bc954754, I've done a pass over the schedules and am putting this PR back in review.

bdach commented:
Test failures look relevant :(

smoogipoo committed:
Restore flow invalidation optimisation

bdach commented on osu.Game/Online/Rooms/MultiplayerPlaylistItem.cs:
```diff
@@ -82,5 +82,24 @@ public MultiplayerPlaylistItem(PlaylistItem item)
+            return ID == other.ID
+                   && OwnerID == other.OwnerID
+                   && BeatmapID == other.BeatmapID
+                   && BeatmapChecksum.Equals(other.BeatmapChecksum, StringComparison.Ordinal)
```
```suggestion
                   && BeatmapChecksum.Equals(other.BeatmapChecksum, StringComparison.OrdinalIgnoreCase)
```

for safety, probably

bdach commented on osu.Game/Online/Rooms/MultiplayerPlaylistItem.cs:
```diff
@@ -82,5 +82,24 @@ public MultiplayerPlaylistItem(PlaylistItem item)
             Freestyle = item.Freestyle;
         }
+
+        public bool Equals(MultiplayerPlaylistItem? other)
```
Where is this equality member used? Rider is only showing me `MultiplayerSpectateButton`; is that it? If it is, can this be removed and a local `ID` equality check be used in that usage site or something?

I'm skittish about this equality check existing because it includes stuff like `StarRating` and `Expired` which is highly incidental.

bdach commented on osu.Game/Screens/OnlinePlay/Match/DrawablePlaylistsRoom.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        public new required Bindable<PlaylistItem?> SelectedItem
+        {
+            get => selectedItem;
```
```suggestion
            get => selectedItem.Current;
```

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/IMultiplayerMatchScreen.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        Room Room { get; }
+
+        bool IsCurrentScreen();
+
+        void Push(IScreen screen);
```
these two existing here is highly weird, would expect them to be in `IScreen` and then `IMultiplayerMatchScreen : IScreen`

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/MatchStartControl.cs:
```diff
@@ -214,8 +206,7 @@ private void updateState()
 
             readyButton.Enabled.Value = countdownButton.Enabled.Value =
                 client.Room.State != MultiplayerRoomState.Closed
-                && SelectedItem.Value?.ID == client.Room.Settings.PlaylistItemId
```
Why is this check gone?

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerSpectateButton.cs:
```diff
@@ -121,11 +116,7 @@ private void updateState()
         {
-            PlaylistItem? item = SelectedItem.Value;
-
-            downloadCheckCancellation?.Cancel();
```
Why was this moved further down after all of the early returns?

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerHistoryList.cs:
```diff
@@ -15,16 +20,82 @@ namespace osu.Game.Screens.OnlinePlay.Multiplayer.Match.Playlist
     /// </summary>
     public partial class MultiplayerHistoryList : DrawableRoomPlaylist
     {
+        public new Bindable<PlaylistItem?> SelectedItem => throw new NotSupportedException();
```
Huh? What is this about?

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerHistoryList.cs:
```diff
@@ -15,16 +20,82 @@ namespace osu.Game.Screens.OnlinePlay.Multiplayer.Match.Playlist
+
+        private void onRoomUpdated() => Scheduler.AddOnce(() =>
+        {
+            if (client.Room == null)
+            {
+                Items.Clear();
+                firstPopulation = true;
+            }
+            else if (firstPopulation)
+            {
+                foreach (var item in client.Room.Playlist)
+                    onItemAdded(item);
+                firstPopulation = false;
+            }
```
Can `MultiplayerClient` not fire `ItemAdded` on initial join or something to avoid playing these `firstPopulation` games here? Because I see that at least 2 components are playing the same boolean flag games and at that point it feels rather icky.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerUserModDisplay.cs:
```diff
@@ -0,0 +1,61 @@
+        [Resolved]
+        private RulesetStore rulesets { get; set; } = null!;
+
+        public new Bindable<IReadOnlyList<Mod>> Current
+        {
+            get => throw new NotSupportedException("Cannot access the current mods of a multiplayer mods display.");
+            set => throw new NotSupportedException("Cannot access the current mods of a multiplayer mods display.");
+        }
```
- does this even matter?
- at this point make this a `CompositeDrawable` with a `ModDisplay` nested rather than subclass or something

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerUserModSelectOverlay.cs:
```diff
@@ -0,0 +1,229 @@
+        /// <returns>
+        /// The token, which should be disposed in order to restore the server's authority.
+        /// </returns>
+        private IDisposable createClientAuthorityToken()
```
i'm gonna need an in-depth explanation of what's going on here because my initial reactions are "what on earth is this" followed by "why though i don't recall seeing something like this ever before in multiplayer code" and a complete brain shutdown and the one sentence about it in the OP isn't helping

bdach reviewed (commented):
i did a pass i guess

to be quite blunt, one of the worst reviewing experiences in a *long* time, i kinda don't believe that this couldn't have been split off more. the amount of changes is overwhelming me, and i posted 12 comments already despite probably giving this like 10% of the scrutiny it actually deserves.

not even sure how to process the 1k line diffstat on `MultiplayerMatchSubScreen` alone other than i guess like empirical testing and going 'looks good :shrug:' if i can't break it

smoogipoo commented:
Point well taken. Let's see how it goes with splitting it up.
